Add requirements for how the Trace API should behave in the absence of an SDK (#719)
open-telemetry/opentelemetry-specification: specification/trace/api.md

| `PRODUCER` | | yes | | maybe |
| `CONSUMER` | | yes | maybe | |
| `INTERNAL` | | | | |

## Behavior of the API in the absence of an installed SDK

> In general, in the absence of an installed SDK, the Trace API is intended to a "no-op" API.

Reviewer: intended to be?

Author: well, I thought so. :) I'll remove the wishy-washy language.

> This means that operations on a Tracer should have no side effects and do nothing. However, there

Reviewer 2: operations on Span as well.

Author: fixed.

> is one important exception to this general rule, and that is related to propagation of a SpanContext.

Reviewer: This part will probably change, based on #721

Reviewer 3: You should be able to install a W3C propagator into the propagation layer (which is supposed to be an artifact separate from API) even without initializing the SDK, right? So I don't think this part changes?

Author: I agree. I don't think this changes. IF there is a propagator installed in the propagation APIs, THEN, this requires the API to propagate any valid spans that are specified as parent spans, even if no real child-span is created.

> Several cases must be considered when a new Span is requested to be created, especially in relation to the

Reviewer 2: Instead of "several", why not "The following"

Author: Yes. good change. will make it so.

Author: done

requested parent SpanContext:

* A valid SpanContext is specified as the parent of the new Span: The API MUST treat this parent context as the

> context for the newly created Span. This means that a SpanContext that has been provided by a configured Propagator

Reviewer 2: Bit confused here.. This section is talking about the scenario when there is *no* Sdk installed. In that scenario, should we still take and propagate the context? Is this propagation limited to in-proc spans only or out-of-proc as well?

Author: So, yes, that's the point of this PR. :) Since you can install a Propagator without an SDK installed, it became important to explain how that would work. If a propagator (let's say, the W3C Trace Context propagator) is installed, this specifies that that API should still propagate incoming spans, transparently. 

It's the "minimum" requirement from the W3C spec for forwarding a trace:  https://www.w3.org/TR/trace-context/#design-overview

Reviewer 4: >It's the "minimum" requirement from the W3C spec for forwarding a trace

If you have no API and no SDK you are not a participant in the trace and thus not accountable to minimum behaviors.

Author: But, if you have propagators attached to the API, shouldn't you be?

> will be propagated through the resulting call stack, but that no new `SpanContext`s will be created.

Reviewer 3: > propagated through the resulting call stack

Not sure what call stacks have to do with this? From where to where should the context be propagated?

```suggestion
will be propagated through to any child span, but that no new `SpanContext`s will be created.
```

Maybe this was meant?

Reviewer 3: I think this needs to be clarified.

Reviewer 2: propagate in-proc and across process boundaries ?

> * No valid `SpanContext` is specified as the parent of the new `Span`: The API MUST create an invalid `Span` for use
> by the API caller. This means that both the `TraceID` and the `SpanID` should be invalid.

Reviewer 3: ```suggestion
* No `SpanContext` is specified as the parent of the new `Span`:
  The API MUST behave as if a `SpanContext` with all-zero `TraceID` and the `SpanID` and empty `TraceState` was passed.
```

I think we should avoid the term "invalid" when possible because it is not clearly defined. Also I think passing a SpanContext with invalid trace ID and a non-empty span ID could be propagated through without requiring to special-case it in the no-op implementation.

Author: I'm just trying to be clear here that you shouldn't generate "fake" span or trace ids for the Span that is returned by the API. They should be fully non-valid (both SpanID and TraceID are all zeroes). I'll add that language.
